File: packages/quant-core/quant_core/ticker_identity.py
"""Canonical ticker identity — the Python twin of the TS `@trader/ticker-identity` adapter.

The platform's single source of truth for "which instrument" is the BARE exchange symbol plus its
listing market — never the broker's concatenated form. A US Alphabet share is
`TickerIdentity('GOOGL', 'US')`; a London Shell share is `TickerIdentity('SHEL', 'LSE')`. Storage
(Mongo fields, Timescale columns, Redis keys) carries `symbol` + `market` separately; the Trading212
`_US_EQ` / `l_EQ` form is produced ONLY by the adapter below, at the broker boundary.

WHY a Python mirror (not a single TS definition): the harvester, fundamentals-api, and the
strategy/backtest hosts are Python and never call the TS package. This module is the byte-for-byte
twin of `packages/ticker-identity/src/adapter.ts` so the two languages cannot drift on the suffix
rules, the currency map, or the FB→META rename. Any change to the broker representation must edit
BOTH adapters (the parity test pins identical fixtures → identical outputs).

`Market` excludes `'OTHER'` deliberately: only US + LSE listings are tradable on the platform, so an
unrecognised suffix is a parse failure (the caller's problem), not a third enum member that would
silently flow into storage and ranking. The legacy `'UK'/'OTHER'` vocabulary of
`quant_core.fundamentals.contract.market_of` is preserved by a thin shim there, which routes through
this adapter and maps `'LSE'→'UK'` / a parse failure → `'OTHER'` (so every existing caller keeps its
byte-identical value).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

# The quote currency vocabulary, mirroring the TS `Currency = 'GBP' | 'USD'`. The account base is
# GBP; the only two tradable markets quote in GBP (LSE) or USD (US).
Currency = Literal["GBP", "USD"]

# `market ∈ {'US','LSE'}` mirrors `instrument_registry.market`; `OTHER` is rejected from the
# tradable universe (an unrecognised suffix throws rather than becoming a third member).
Market = Literal["US", "LSE"]

# T212 represents a listing by appending an exchange tag to the bare symbol:
#   US  →  `<symbol>_US_EQ`   (e.g. GOOGL → GOOGL_US_EQ)
#   LSE →  `<symbol>l_EQ`     (lowercase 'l' joined to the symbol, e.g. SHEL → SHELl_EQ)
# These two literals are the entire knowledge of the broker representation in the Python codebase —
# every other call site routes through `to_t212` / `from_t212`.
_US_SUFFIX = "_US_EQ"
_LSE_SUFFIX = "l_EQ"

# Legacy-rename table: a symbol the broker's catalog still lists under its pre-rebrand name maps to
# the canonical post-rebrand symbol. Seeded from the scattered `SYMBOL_RENAMES` maps — Facebook →
# Meta. Keyed by `(market, from_symbol)` because a rebrand is listing-specific: FB→META is a US
# event, so it must not silently rewrite a same-named symbol on another market.
_RENAMES: dict[Market, dict[str, str]] = {
    "US": {"FB": "META"},
    "LSE": {},
}

# Currency is a pure function of the listing market — this replaces the duplicated
# `currencyOfTicker` / `inferCurrency` suffix sniffers so a single rule governs every side.
_CURRENCY_BY_MARKET: dict[Market, Currency] = {
    "US": "USD",
    "LSE": "GBP",
}


@dataclass(frozen=True)
class TickerIdentity:
    """The bare exchange symbol (`GOOGL`, `SHEL`) plus its listing market — the source of truth.

    `frozen` so an identity is a value object (hashable, usable as a dict key) and cannot be mutated
    after the adapter constructs it; the TS interface is `readonly` for the same reason.
    """

    symbol: str
    market: Market


def _require_symbol(symbol: str) -> str:
    """A bare symbol must be a non-empty token with no broker suffix already attached — guards
    against `to_t212(TickerIdentity('', …))` emitting a suffix-only string and against
    double-encoding an already-T212 value back through the adapter."""
    s = (symbol or "").strip()
    if len(s) == 0:
        raise ValueError("[ticker-identity] empty symbol")
    return s


class Trading212TickerAdapter:
    """The only code in the Python services that produces or parses the Trading212 `_US_EQ` /
    `l_EQ` form, derives currency from a listing, or owns the legacy symbol rename. Everything
    upstream works in `TickerIdentity` and converts at the broker boundary alone.
    """

    def to_t212(self, ident: TickerIdentity) -> str:
        """Identity → the broker's ticker string. `{GOOGL, US} → 'GOOGL_US_EQ'`;
        `{SHEL, LSE} → 'SHELl_EQ'`. Inverse of :meth:`from_t212`.

        Throws on a market outside `US|LSE` (symmetric with :meth:`from_t212`'s rejection): a
        `market` hydrated from storage can smuggle an out-of-type value, so fail loudly rather than
        broker-sending a malformed string.
        """
        symbol = _require_symbol(ident.symbol)
        if ident.market == "US":
            return f"{symbol}{_US_SUFFIX}"
        if ident.market == "LSE":
            return f"{symbol}{_LSE_SUFFIX}"
        raise ValueError(f"[ticker-identity] unsupported market: {ident.market!r}")

    def from_t212(self, t212: str) -> TickerIdentity:
        """The broker's ticker string → identity. The ONLY suffix parser in the Python codebase,
        and a strict inverse of :meth:`to_t212`: `'GOOGL_US_EQ' → {GOOGL, US}`,
        `'SHELl_EQ' → {SHEL, LSE}`. A string that is not a recognised US/LSE equity form (an
        unsupported market, CFD, or malformed input) throws — it is not silently coerced to a third
        market, because only US + LSE are tradable here.
        """
        raw = (t212 or "").strip()
        # US is unambiguous: the explicit `_US_EQ` tail. Strip it for the bare symbol.
        if raw.endswith(_US_SUFFIX):
            symbol = raw[: -len(_US_SUFFIX)]
            if len(symbol) == 0:
                raise ValueError(f"[ticker-identity] malformed T212 ticker (empty symbol): {t212}")
            return TickerIdentity(symbol=symbol, market="US")
        # LSE is `<symbol>l_EQ`: the `l` belongs to the suffix, so the bare symbol is what precedes
        # `l_EQ` (and must be non-empty — `l_EQ` alone has no symbol).
        if raw.endswith(_LSE_SUFFIX):
            symbol = raw[: -len(_LSE_SUFFIX)]
            if len(symbol) == 0:
                raise ValueError(f"[ticker-identity] malformed T212 ticker (empty symbol): {t212}")
            return TickerIdentity(symbol=symbol, market="LSE")
        raise ValueError(f"[ticker-identity] unrecognised T212 ticker (not a US/LSE equity): {t212}")

    def currency_of(self, ident: TickerIdentity) -> Currency:
        """Listing market → quote currency. `US → 'USD'`, `LSE → 'GBP'`."""
        return _CURRENCY_BY_MARKET[ident.market]

    def apply_rename(self, ident: TickerIdentity) -> TickerIdentity:
        """Apply a legacy symbol rename (e.g. US `FB → META`) market-aware. Returns a new identity
        with the canonical symbol when a rename applies, otherwise the input unchanged. Replaces the
        scattered `SYMBOL_RENAMES` lookups.
        """
        symbol = _require_symbol(ident.symbol)
        renamed = _RENAMES[ident.market].get(symbol)
        if renamed is None or renamed == symbol:
            return ident
        return TickerIdentity(symbol=renamed, market=ident.market)

File: services/frontend-web/portal/app/lib/ticker-identity.ts
// Portal-local mirror of `@trader/ticker-identity` (packages/ticker-identity/src/adapter.ts). The
// portal is a standalone Next.js app OUTSIDE the pnpm workspace, so it cannot import the workspace
// package — this is the documented "local mirror" pattern (see the `Mirror of @trader/shared-types`
// notes elsewhere in the portal). Keep the suffix rule + the FB→META rename in lockstep with the
// upstream adapter; this is the ONLY place the portal knows the `_US_EQ` / `l_EQ` broker form.
//
// Why the portal needs it (epic pit-fundamentals-lake-rearchitecture, Task 21): the bare ticker
// `(symbol, market)` is the platform source of truth. The universe-overrides API still echoes the
// legacy T212 form on the wire (Task 16b/18 left that for Task 21 to remove), so the portal parses it
// to the bare identity for display + posts the bare `{symbol, market}` form back (the backend PUT has
// accepted `{symbol, market}` objects since Task 18). Rendering is bare `symbol`; the broker suffix
// never reaches the operator.

export type Market = 'US' | 'LSE'

export interface TickerIdentity {
  symbol: string
  market: Market
}

const US_SUFFIX = '_US_EQ'
const LSE_SUFFIX = 'l_EQ'

// Legacy-rename table — keyed by (market, fromSymbol) because a rebrand is listing-specific. Seeded
// from the upstream adapter's RENAMES (US Facebook → Meta). Kept narrow on purpose.
const RENAMES: Record<Market, Record<string, string>> = {
  US: { FB: 'META' },
  LSE: {},
}

/**
 * The broker's ticker string → bare identity, the strict inverse of {@link toT212}:
 * `'GOOGL_US_EQ' → { GOOGL, US }`, `'SHELl_EQ' → { SHEL, LSE }`.
 *
 * TOLERANT by design (unlike the upstream adapter, which throws): the portal renders whatever the
 * active universe contains, and a non-US/LSE form (an `OTHER` suffix, a CFD, a malformed value, or —
 * post-cutover — an already-bare symbol) must not crash a list render. A string that isn't a
 * recognised US/LSE equity form returns `null`; the caller falls back to showing the raw value.
 */
export function fromT212(t212: string | null | undefined): TickerIdentity | null {
  const raw = (t212 ?? '').trim()
  if (raw.endsWith(US_SUFFIX)) {
    const symbol = raw.slice(0, -US_SUFFIX.length)
    return symbol.length === 0 ? null : { symbol, market: 'US' }
  }
  if (raw.endsWith(LSE_SUFFIX)) {
    const symbol = raw.slice(0, -LSE_SUFFIX.length)
    return symbol.length === 0 ? null : { symbol, market: 'LSE' }
  }
  return null
}

/**
 * Bare identity → the broker's ticker string. `{ GOOGL, US } → 'GOOGL_US_EQ'`;
 * `{ SHEL, LSE } → 'SHELl_EQ'`. The portal only needs this where it must echo a legacy-shaped value
 * back to an endpoint that hasn't migrated; the universe PUT accepts the bare object directly, so
 * this is mostly here for symmetry + tests.
 */
export function toT212(id: TickerIdentity): string {
  const symbol = id.symbol.trim()
  if (symbol.length === 0) throw new Error('[ticker-identity] empty symbol')
  return id.market === 'US' ? `${symbol}${US_SUFFIX}` : `${symbol}${LSE_SUFFIX}`
}

/** Apply the market-aware legacy rename (US `FB → META`); returns the input unchanged when none applies. */
export function applyRename(id: TickerIdentity): TickerIdentity {
  const renamed = RENAMES[id.market][id.symbol.trim().toUpperCase()]
  if (renamed === undefined || renamed === id.symbol) return id
  return { symbol: renamed, market: id.market }
}

/**
 * Normalise a forced-add as the operator types it into the bare identity the UI stores + posts.
 * Accepts a bare `'GOOGL'` (market from the selector, default US), an already-bare-but-explicit
 * value, or a pasted legacy T212 string (`'AAPL_US_EQ'` / `'SGLNl_EQ'` — the market in the suffix
 * wins over the selector). Returns `null` on an empty symbol so the caller can ignore the entry.
 *
 * The bare symbol is upper-cased + the rename applied so the stored identity is canonical (META, not
 * FB) and matches the backend's own `resolveForcedAdd` normalisation. The market selector is the
 * fallback ONLY when the typed value carries no listing of its own.
 */
export function parseForcedAdd(raw: string, marketHint: Market): TickerIdentity | null {
  const s = raw.trim()
  if (s === '') return null
  // A pasted legacy T212 string carries its own listing — the suffix market wins over the selector.
  // Try the canonical-case string first (handles `SGLNl_EQ` whose lowercase `l` is load-bearing),
  // then an upper-cased retry to catch a lower-cased US form (`aapl_us_eq`).
  for (const candidate of s === s.toUpperCase() ? [s] : [s, s.toUpperCase()]) {
    const parsed = fromT212(candidate)
    if (parsed) return applyRename({ symbol: parsed.symbol.toUpperCase(), market: parsed.market })
  }
  // A bare symbol — use the selector's market.
  return applyRename({ symbol: s.toUpperCase(), market: marketHint })
}

/** Stable de-dup key for an identity (`SYMBOL|MARKET`). */
export function identityKey(id: TickerIdentity): string {
  return `${id.symbol.toUpperCase()}|${id.market}`
}
